### PR TITLE
Add a FLATTEN_RECORD_LIST formula

### LINKED ISSUE #1130: Lists of RecordSets cannot be typed as Reference List
### Problem

Say in table Carts there's a column $Cart of type Reference List, pointing to a Foods table

In a third table (say, Customers), we want to list all Foods that are in a customer's cart.

`Carts.lookupRecords(customer=$id).Cart` returns a "list of lists" (`[Foods[[1, 2, 3]], Foods[[1]]]`) which is not handled properly by Grist when trying to type it as a Reference List

A fix is to return a flattened list using a list comprehension `[el for val in Carts.lookupRecords(customer=$id).Cart for el in val]`

An example document can be found here : https://docs.getgrist.com/bAjQhzJRgBJB/Issue-1130-Example/p/1

### Proposed solution

Option 1 : Implement a `FLATTEN_RECORDS_LIST` function that transforms a list of lists of records into a list of records. We should offer a deduplicate option to this function.

Option 2 : Make Grist handle those list of lists of records as a valid input for the Reference List type

WDYT ?

### PR DESCRIPTION
**Pull request by hexaltation**
## Context

Formulas `Carts.lookupRecords(customer=$id).Cart` returns a "list of lists/RecordSet" ([Foods[[1, 2, 3]], Foods[[1]]]) which is not handled properly by Grist when trying to type it as a Reference List.

A fix is to return a flattened list using a list comprehension [el for val in Carts.lookupRecords(customer=$id).Cart for el in val]

## Proposed solution

Add a `FLATTEN_RECORD_LIST(RecordSet, dedup=False)` formula that returns a much more manageable list of `Records`

## Related issues

Fixes #1130

## Has this been tested?

- [ ] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [x] 🙋 no, because I need help

I'm not sure of the best way to test this. Creating a .grist file that verify the formula returns the good result in nbrowser tests or python tests inside the docstring, but how create a proper RecordSet to do unit tests ?

Thanks for your help.

### CODE REVIEW

File: sandbox/grist/functions/lookup.py [hunk 1]
```diff
@@ -26,6 +26,18 @@ def COLUMNS(range):
   """Returns the number of columns in a specified array or range."""
   raise NotImplementedError()
 
+def FLATTEN_RECORDS_LIST(list_of_records, dedup=False):
+    """Returns a flat list of records"""
+    if type(list_of_records) is not list:
+        raise TypeError("Unexpected Argument, waiting of a list of Records")
```
Comment (fflorent) [inline sandbox/grist/functions/lookup.py:32]: wouldn't `waiting for a list of Records` more English?

File: sandbox/grist/functions/lookup.py [hunk 2]
```diff
@@ -26,6 +26,18 @@ def COLUMNS(range):
...
+    if len(list_of_records) == 0:
+        return list_of_records
+    flatten_list = [el for val in list_of_records for el in val]
+    el_type = {el.__class__.__name__ for el in flatten_list}
+    if list(el_type)[0] != 'Record':
```
Comment (fflorent) [inline sandbox/grist/functions/lookup.py:37]: Why do you have to check the first element only?

Wouldn't `'Record' in el_type` be more robust?

**Review by fflorent — CHANGES_REQUESTED** [review]
Thanks @hexaltation! Few remarks from my side.

**Comment by berhalak** [comment]
Closing this one. I'll create new one with option 2 from the issue.

**Pull request closed without merging**